src/Symfony/Component/HttpKernel/Profiler/ProfileStack.php

```diff
@@ -0,0 +1,50 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel\Profiler;
+
+use Symfony\Component\HttpFoundation\Request;
+
+/**
+ * @internal
+ */
+final class ProfileStack
+{
+    /**
+     * @var \SplObjectStorage
+     */
+    private $profiles;
+
+    public function __construct()
+    {
+        $this->reset();
+    }
+
+    public function has(Request $request): bool
+    {
+        return isset($this->profiles[$request]);
+    }
+
+    public function get(Request $request): Profile
+    {
+        return $this->profiles[$request];
```

Comment:
the implementation does not deal with cases where the request does not have a profile, which would break with a TypeError (which would then be considered as a bug in Symfony by users, and rightfully so). You should check this case and throw a meaningful exception.

Reply:
The class is `@internal`. It's our responsibility to use `get()` after `has()` (that's what is done). Isn't it enough? 🤔

Reply:
I would still make the code report a meaningful error if you misuse it rather than a TypeError. Especially given that it does not either document the responsibility it expects from us (and so future contributors may not know it before using it).
I'm fine with not adding doc explaining that if the failure mode is clear. But a TypeError saying we cannot return `null` or a failure saying that the item does not exist in the SplObjectStorage (I don't remember how SplObjectStorage deals with that) is not a clear failure mode.